lib/models/patch/file-patch.js

```diff
@@ -0,0 +1,283 @@
+  getNoNewlineRanges() {
+    const hunks = this.getHunks();
+    const lastHunk = hunks[hunks.length - 1];
+    if (!lastHunk) {
+      return [];
+    }
+
+    const range = lastHunk.getNoNewlineRange();
+    if (!range) {
+      return [];
+    }
+
+    return [range];
+  }
+
+  adoptBufferFrom(prevFilePatch) {
+    this.getPatch().adoptBufferFrom(prevFilePatch.getPatch());
+  }
+
+  didChangeExecutableMode() {
+    if (!this.oldFile.isPresent() || !this.newFile.isPresent()) {
+      return false;
+    }
+
+    return this.oldFile.isExecutable() && !this.newFile.isExecutable() ||
+      !this.oldFile.isExecutable() && this.newFile.isExecutable();
+  }
+
+  hasSymlink() {
+    return Boolean(this.getOldFile().getSymlink() || this.getNewFile().getSymlink());
+  }
+
+  hasTypechange() {
+    if (!this.oldFile.isPresent() || !this.newFile.isPresent()) {
+      return false;
+    }
+
+    return this.oldFile.isSymlink() && !this.newFile.isSymlink() ||
+      !this.oldFile.isSymlink() && this.newFile.isSymlink();
+  }
+
+  getPath() {
+    return this.getOldPath() || this.getNewPath();
+  }
+
+  getStatus() {
+    return this.getPatch().getStatus();
+  }
+
+  getHunks() {
+    return this.getPatch().getHunks();
+  }
+
+  clone(opts = {}) {
+    return new this.constructor(
+      opts.oldFile !== undefined ? opts.oldFile : this.oldFile,
+      opts.newFile !== undefined ? opts.newFile : this.newFile,
+      opts.patch !== undefined ? opts.patch : this.patch,
+    );
+  }
+
+  getStagePatchForLines(selectedLineSet) {
+    if (this.patch.getChangedLineCount() === selectedLineSet.size) {
+      if (this.hasTypechange() && this.getStatus() === 'deleted') {
+        // handle special case when symlink is created where a file was deleted. In order to stage the file deletion,
+        // we must ensure that the created file patch has no new file
+        return this.clone({newFile: nullFile});
+      } else {
+        return this;
+      }
+    } else {
+      const patch = this.patch.getStagePatchForLines(selectedLineSet);
+      if (this.getStatus() === 'deleted') {
+        // Populate newFile
+        return this.clone({newFile: this.getOldFile(), patch});
+      } else {
+        return this.clone({patch});
+      }
+    }
+  }
+
+  getStagePatchForHunk(selectedHunk) {
+    return this.getStagePatchForLines(new Set(selectedHunk.getBufferRows()));
+  }
+
+  getUnstagePatchForLines(selectedLineSet) {
+    if (this.patch.getChangedLineCount() === selectedLineSet.size) {
+      const patch = this.patch.getFullUnstagedPatch();
+      if (this.hasTypechange() && this.getStatus() === 'added') {
+        // handle special case when a file was created after a symlink was deleted.
+        // In order to unstage the file creation, we must ensure that the unstage patch has no new file,
+        // so when the patch is applied to the index, there file will be removed from the index.
+        return this.clone({oldFile: nullFile, patch});
```

Comment:
There's a bug hiding here... 

Repro steps:
1. Delete symlink file
1. Create new file with same name as deleted symlink file
1. Stage whole file
1. Select all lines in file and click "Unstage Selected Line" button

The diff produced is (`this.clone({oldFile: nullFile, patch}).toString()`)

```
"diff --git a/symlink.txt b/symlink.txt
deleted file mode 120000
--- a/symlink.txt
+++ /dev/null
@@ -1 +0,0 @@
-./regular-file.txt
\ No newline at end of file
"
```

That hunk should contain deleted lines of the newly added file, not the symlink contents...

And the following error is thrown:

```
git-shell-out-strategy.js? [sm]:32 Uncaught (in promise) GitError: git apply --cached - in /Users/kuychaco/src/repo-symlinks exited with code 1
stdout: 
stderr: error: symlink.txt: wrong type

    at new GitError (/Users/kuychaco/github/github/lib/git-shell-out-strategy.js:34:18)
    at Promise (/Users/kuychaco/github/github/lib/git-shell-out-strategy.js:317:23)
    at <anonymous>
```